[HMA] Change matching API to return more metadata

aokolish opened the pull request:
Summary
---------

This PR changes the API responses for the GET and POST `/m/lookup` APIs based on issue #1681. The motivation being - users of the API need to know the content IDs so that they can turn content IDs off if they're causing false positives. Also, callers may find it useful to log / analyze matching distance as well.

**API response before:** (API just returns list on bank names with matches)
```
["FOO_BANK"]
```

**API response w/this change:** (API returns a dict of bank names with matches + details about each match)
```
{
    "FOO_BANK": [
        {
            "content_id": 16,
            "distance": "13"
        }
    ]
}
```

Note: I modified tests that were calling `/raw_lookup` because there were no tests for the `/lookup` API. I also did that in part because I think we could remove `raw_lookup` once this is merged since the main benefit of that API to me was that it returned the content IDs.

@Dcallies I'm opening this as a draft because I couldn't figure out how to get `3P_exchange_type` and `3P_Records` from the issue. If it's not too hard to add, we can do it now. Otherwise, this is good enough to be valuable to me as-is.

Test Plan
---------

Unit tests + testing in my local HMA dev instance

Here's an example from my dev instance:
```shell
$ curl -s "http://localhost:5000/m/lookup?signal_type=pdq&signal=cccccccc33133333eccecccc33313333cccccccc33313333cce6cccc33933333" | jq
{
  "FOO_BANK": [
    {
      "content_id": 81842,
      "distance": 20
    },
    {
      "content_id": 82244,
      "distance": 10
    },
    {
      "content_id": 82519,
      "distance": 18
    },
    {
      "content_id": 82850,
      "distance": 18
    },
    {
      "content_id": 82870,
      "distance": 18
    },
    {
      "content_id": 82883,
      "distance": 24
    },
    {
      "content_id": 91618,
      "distance": 14
    }
  ]
}
```

aokolish commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
+
+        match = results_by_content_id.get(content.id)
+        results[content.bank.name].append(
+            # is it possible to return the content_signal.signal_val here?
```
@Dcallies I also wanted to return the hash from the bank but couldn't figure that one out. I think each `bank_content` has many `content_signal`s, right? Maybe it's hard to get it here.

Dcallies replied:
It is possible, but it will take more round trips to the DB. I think adding a parameter to request extra information that we can't get "for free", but are useful for debugging makes sense. 

It looks like we didn't expose ContentSignal - it just needs a simple Crud that takes  (bank_content_id), and we'll probably want a multi-lookup so you can use it this context.

aokolish replied:
Yeah, I changed my mind and am thinking we don't need this. It's more of a nice to have and most of the time we won't need it. I can get it via other techniques, if / when I need it.

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
+    results = defaultdict(list)
+    for content in enabled_content:
+        if content.bank.name not in enabled_banks:
+            next
```
blocking: I don't think `next` is a keyword in python is it? `continue`?

aokolish replied:
whoops, I'm more fluent in ruby...

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
     current_app.logger.debug("performing lookup")
-    raw_results = lookup_signal(signal, signal_type_name)
+    results_by_content_id = {
+        r.metadata: r for r in query_index(signal, signal_type_name)
```
nit: might want to add `results_by_bank_content_id` to differentiate from a future where we might have on-platform content ids as well

aokolish replied:
Did you mean to rename the variable? I didn't totally understand your comment, but that's what I did.

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
-    current_app.logger.debug(raw_results)
-    contents = storage.bank_content_get(raw_results)
-    enabled = [c for c in contents if c.enabled]
+    contents = storage.bank_content_get(results_by_content_id.keys())
```
ignorable: I think this function takes a traversable, and the traverse for dicts is the keys

aokolish replied:
changed

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
+    }
     current_app.logger.debug(
-        "lookup matches %d banks (%d enabled)", len(banks), len(enabled_banks)
+        "lookup matches %d banks (%d enabled_content)", len(banks), len(enabled_banks)
```
blocking: this is enabled banks, not enabled content

aokolish replied:
fixed

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/tests/test_e2e_workflow.py:
```diff
@@ -37,37 +37,13 @@ def test_raw_hash_add_to_match_no_distance(app: Flask, client: FlaskClient):
+    resp = client.get(f"/m/lookup?signal_type=pdq&signal={hashes[-1]}")
     assert resp.status_code == 200
-    assert resp.json == {"matches": [16]}
+    assert resp.json == {"TEST_BANK": [{"content_id": 16, "distance": 0}]}
```
<3

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +248,40 @@ def lookup_post():
 
 def lookup(signal, signal_type_name):
```
blocking: Make sure to update the documentation on the lookup functions

Dcallies replied:
close enough

Dcallies reviewed (changes requested):
Overall change looks good, and we can get the content id and distance "for free" (no more additional database calls), so I think it makes sense to return by default.

For additional data (3P Id, hash string, etc), maybe we can add those as optional parameters, since they take extra calls to fetch.

We may need to add a new storage interface to allow getting the signal strings. Seems like an oversight in the Bank interface.

Let me know if you want to tag team this - I can help on the lower parts of the stack

aokolish commented:
@Dcallies I think I addressed all your feedback. This version is adding value for me (enables me to log data regarding the matches we're finding) so I'd like to merge it without any additional data when you're ok with it. Meaning, I don't need the hash values or 3P data...just the content IDs and distance are the main things I'll be needing for now.

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/blueprints/matching.py:
```diff
@@ -247,27 +252,39 @@ def lookup_post():
+        match = results_by_bank_content_id.get(content.id)
+        results[content.bank.name].append(
+            {
+                "content_id": content.id,
```
blocking: I missed this version of it, but I forsee the need for a "content id" down the line, and so we should make this clear it's the `bank_content_id`

Dcallies commented on hasher-matcher-actioner/src/OpenMediaMatch/tests/test_e2e_workflow.py:
```diff
@@ -37,37 +37,13 @@ def test_raw_hash_add_to_match_no_distance(app: Flask, client: FlaskClient):
+    # Test lookup with no match
     resp = client.get(
-        f"/m/raw_lookup?signal_type=pdq&include_distance=true&signal={hashes[-1]}"
+        "/m/lookup?signal_type=pdq&signal=5555555555555555555555555555555555555555555555555555555555555555"
```
You can save some space if you want by doing "5" * 64

```
"/m/lookup?signal_type=pdq&signal={}".format("5" * 64)
```

Dcallies reviewed (approved):
Looks good, one last rename. Let me know what you want to do about the CLA.

This will be a backwards incompatible change for current users of the API, but I think we should break this for the structured version.

Dcallies commented:
I'll run the formatter on your code afterwards
